ouster_client/src/client.cpp

```diff
@@ -154,6 +154,90 @@ SOCKET udp_data_socket(int port) {
     return SOCKET_ERROR;
 }
 
+
+SOCKET mtp_data_socket(int port, const std::string& udp_dest_host = "",
+                       const std::string& mtp_dest_host = "") {
+    struct addrinfo hints, *info_start, *ai;
+
+    memset(&hints, 0, sizeof hints);
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_DGRAM;
+    hints.ai_flags = AI_PASSIVE;
+
+    auto port_s = std::to_string(port);
+
+    int ret = getaddrinfo(NULL, port_s.c_str(), &hints, &info_start);
+    if (ret != 0) {
+        logger().error("mtp getaddrinfo(): {}", gai_strerror(ret));
+        return SOCKET_ERROR;
+    }
+    if (info_start == NULL) {
+        logger().error("mtp getaddrinfo(): empty result");
+        return SOCKET_ERROR;
+    }
+
+    for (auto preferred_af : {AF_INET}) { // TODO test with AF_INET6
+        for (ai = info_start; ai != NULL; ai = ai->ai_next) {
+            if (ai->ai_family != preferred_af) continue;
+
+            // choose first addrinfo where bind() succeeds
+            SOCKET sock_fd = 
+                socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
+            if (!impl::socket_valid(sock_fd)) {
+                logger().warn("mtp socket(): {}", impl::socket_get_error());
+                continue;
+            }
+
+            if (impl::socket_set_reuse(sock_fd)) {
+                logger().warn("mtp socket_set_reuse(): {}",
+                              impl::socket_get_error());
+            }
+
+            if (::bind(sock_fd, ai->ai_addr, (socklen_t)ai->ai_addrlen)) {
+                logger().warn("mtp bind(): {}", impl::socket_get_error());
+                impl::socket_close(sock_fd);                
+                continue;
+            }
+
+            // bind() succeeded; join to multicast group on with preferred address
+            // connect only if addresses are not empty
+            if (!udp_dest_host.empty() && !mtp_dest_host.empty()) {
```

Comment:
So that people don't need to specify the interface can we default to INADDR_ANY if the mtp_dest_host is not specified. 
Something like:
```
if (!udp_dest_host.empty()) {
                ip_mreq mreq;
                mreq.imr_multiaddr.s_addr = inet_addr(udp_dest_host.c_str());
                if (!mtp_dest_host.empty()) {
                    mreq.imr_interface.s_addr =
                        inet_addr(mtp_dest_host.c_str());
                } else {
                    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
                }
```

Reply:
@MattC11 The second ` if (!mtp_dest_host.empty()) {` will always be true, did you instead mean to say?
```cpp
if (!udp_dest_host.empty()) {
    ip_mreq mreq;
    mreq.imr_multiaddr.s_addr = inet_addr(udp_dest_host.c_str());
    if (!mtp_dest_host.empty()) {
        mreq.imr_interface.s_addr =
        inet_addr(mtp_dest_host.c_str());
    } // <- THIS BRACE WAS PROBABLY MISSING      
} else {
    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
}
```

Reply:
I think I had it right, udp_dest_host needs to be set. Note that I removed the second part of the if on line 204. That is the multicast ip address that the sensor is sending to so if it doesn't exist we don't have anything to subscribe to. And actually if it isn't set maybe we should exit the whole function right away?

The check I was trying to add is if the mtp_dest_host (this client machine IP address of one of it's interfaces) is empty in which case we subscribe to any interface.

Reply:
But what if we have several physical interfaces? in that case INADDR_ANY will return 1st iface. For example: our case. We have several Vecow EAC-3000 (Jetson Xavier) that have 5 physical interfaces on our robot. One of them we use to collect  sensor data from Ousters (it is a noticeably traffic if we use 2 OS0-64) and another for sending/receiving ROS communication messages for autonomus driving. 

May be i can try to make INADDR_ANY default option for mtp_dest in launch?

Reply:
Yeah, @MattC11 is right - udp_dest_host needs to be set. And working code looks so:
```suggestion
            // bind() succeeded; join to multicast group on with preferred address
            // connect only if addresses are not empty
            if (!udp_dest_host.empty()) {
                ip_mreq mreq;
                mreq.imr_multiaddr.s_addr = inet_addr(udp_dest_host.c_str());
                if (!mtp_dest_host.empty()) {
                    mreq.imr_interface.s_addr = inet_addr(mtp_dest_host.c_str());
                } else {
                    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
                }

                if (setsockopt(sock_fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, (char*)&mreq,
                            sizeof(mreq))) {
                    logger().warn("mtp setsockopt(): {}", impl::socket_get_error());
                    impl::socket_close(sock_fd);
                    continue;
                }
            }
```

But(!), there is small problem: [here](https://github.com/ShepelIlya/ouster-ros/blob/feature/multicast_ros_driver/src/os_sensor_nodelet.cpp#L300-L305) i use mtp_dest_arg to check if we need to use multicast version of init_client.

I can add default constant  INADDR_ANY in launch like you did with possible values of timestamp mode 
OR
I can add additional rosparam to check if we need multicast version.

Which option do you prefer? Or may be smth else?

Reply:
I'd be inclined to just check if the address range is in the multicast range which will determine which function you call (line 166 of the os_sensor_nodelet.cpp) . A number of ways to do it here: https://stackoverflow.com/questions/22015308/c-function-to-determine-if-ip-address-is-multicast-address
That way you don't need any extra user input.

Reply:
Alternatively, yes putting it to INADDR_ANY in your launch file so that you have to specify mtp_dest_arg it if you want to be multicast sounds reasonable as well.

Reply:
> I'd be inclined to just check if the address range is in the multicast range which will determine which function you call (line 166 of the os_sensor_nodelet.cpp) . A number of ways to do it here: https://stackoverflow.com/questions/22015308/c-function-to-determine-if-ip-address-is-multicast-address That way you don't need any extra user input.

Yeah, thats good point! I`ll do it

Reply:
I think  [this](https://github.com/ouster-lidar/ouster_example/pull/480/commits/6a42349849740291551c11b631c8424fee9732e3) should work, can you pls check it too with your test code @MattC11 ?

Corresponding [commit](https://github.com/ouster-lidar/ouster-ros/pull/57/commits/200c254f2dff242565a0e1a15e4ea70b5f8eec5e) in ouster_ros.

Reply:
I tried it out with my test code and it worked.

Reply:
Nice, than i resolve it